### PR TITLE
totp : rework and add SHA256/512

### PR DESCRIPTION
**Pull request by EmilienCourt**
This Pull Request adds support for `HMAC_SHA224`, `HMAC_SHA256`, `HMAC_SHA384` and `HMAC_SHA512`. The TOTP code has been reworked, and adding new algorithms should be easier.

More precisely, I've :

- reworked the SHA* code to use [mbedtls](https://github.com/Mbed-TLS/mbedtls). My initial goal was just to add SHA256/SHA512 variants to the existing code, but it turns out that the library that is used ([TOTP-MCU](https://github.com/Netthaw/TOTP-MCU)) only has the SHA1 implementation.  In my experience, mbedtls' code is cleaner, easier to understand and more used. I also did not want to introduce a lack of code uniformity (for example, SHA1 being from TOTP-MCU and SHA256/SHA512 from mbedtls)

- reworked the HMAC code, to make it more standard, readable and effective. I've used part of [this project](https://github.com/mygityf/cipher/blob/master/cipher/hmac.c) for the computations

- reworked the TOTP code to take into account those modifications. The supported algorithms are now **SHA1, SHA224, SHA256, SHA384, SHA512** instead of just SHA1

- reworked the `totp_face` and `totp_face_lfs` faces to take into account those modifications

- reworked the examples to test the new algorithms

- fixed a bug in [totp_face_lfs.c](https://github.com/joeycastillo/Sensor-Watch/blob/b7a461d280d06749aeb14af1e8c17cbf3616055b/movement/watch_faces/complication/totp_face_lfs.c#L87) which prevented that face from working correctly

All tests were made using [mintotp](https://github.com/susam/mintotp):
```bash
mintotp 30 6 sha1 <<< JBSWY3DPEHPK3PXP
mintotp 30 6 sha224 <<< JBSWY3DPEHPK3PXP
mintotp 30 6 sha256 <<< JBSWY3DPEHPK3PXP
mintotp 30 6 sha384 <<< JBSWY3DPEHPK3PXP
mintotp 30 6 sha512 <<< JBSWY3DPEHPK3PXP
```

I've tested the code only in the simulator for now, I should be able to test it on the watch soon

Don't hesitate to make comments on the code : I'm not a code nor a crypto expert :)

### CODE REVIEW

File: movement/watch_faces/complication/totp_face_lfs.c [hunk 1]
```diff
@@ -40,6 +40,7 @@ struct totp_record {
     size_t secret_size;
     char label[2];
     uint32_t period;
+    hmac_alg algorithm;
```
Comment (wryun) [inline movement/watch_faces/complication/totp_face_lfs.c:43]: Probably nice to explicitly initialise the algorithm to SHA1 in `init_totp_record` below.

**Review by wryun — COMMENTED** [review]
Nice! Oddly enough, I don't seem to have any TOTP codes that use other algorithms, so this never came up for me.

Haven't looked that deeply at this (or run it), just a quick browse. My main thought is that we might want to make it clearer that this is no longer just a copied in 3rd party library. This might just be amending the README to say what we've done compared to upstream.

File: movement/watch_faces/complication/totp_face_lfs.c (hunk 1 above)
Comment (EmilienCourt) [reply movement/watch_faces/complication/totp_face_lfs.c:43]: You are right :). This was fixed in [ca7806a](https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/ca7806a79b806e1a82373e9c805f2d5b80a33f7c).

**Comment by EmilienCourt** [comment]
> Nice! Oddly enough, I don't seem to have any TOTP codes that use other algorithms, so this never came up for me.
> 
> Haven't looked that deeply at this (or run it), just a quick browse. My main thought is that we might want to make it clearer that this is no longer just a copied in 3rd party library. This might just be amending the README to say what we've done compared to upstream.

Hi !

Thanks for your feedback !

I had indeed missed that README file.

I edited it and also edited the example code, and moved TOTP-MCU to TOTP : [ae6dc3b](https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/ae6dc3b344493f9d38f9f64645568b8c586b3c71).

Still waiting to be able to test this code on the real board.

File: movement/lib/TOTP/sha1.c [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+  memcpy(buffer + SHA1_BLOCK_LENGTH, out, SHA1_DIGEST_LENGTH);
+  mbedtls_sha1(buffer, SHA1_BLOCK_LENGTH + SHA1_DIGEST_LENGTH, out);
+}
+/*
+* Compute TOTP_HMAC_SHA1 using key, key length, text to hash, size of the text
+*/
+uint32_t TOTP_HMAC_SHA1(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n){
+    // STEP 1, get the HMAC-SHA1 hash from counter and key
+    uint8_t hash[SHA1_DIGEST_LENGTH];
+    HMAC_SHA1(key, key_length, in, 8, hash);
```
Comment (joeycastillo) [inline movement/lib/TOTP/sha1.c:382]: It seems here like we're ignoring the text size and hard-coding a length of 8 in the call to `HMAC_SHA1`. Shouldn't this be:

```c
HMAC_SHA1(key, key_length, in, n, hash);
```

I see the same issue in `TOTP_HMAC_SHA256` and `TOTP_HMAC_SHA512`.

File: movement/lib/TOTP/sha256.c [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+*/
+uint32_t TOTP_HMAC_SHA256(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, int is224){
+    int digest_length = SHA256_DIGEST_LENGTH;
+    if (is224 == 1) {
+      digest_length = SHA224_DIGEST_LENGTH;
+    }
+
+    // STEP 1, get the HMAC-SHA256 hash from counter and key
+    uint8_t hash[digest_length];
+    HMAC_SHA256(key, key_length, in, 8, hash, is224);
```
Comment (joeycastillo) [inline movement/lib/TOTP/sha256.c:356]: Same issue as in TOTP_HMAC_SHA1.

File: movement/lib/TOTP/sha512.c [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+*/
+uint32_t TOTP_HMAC_SHA512(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, int is384){
+    int digest_length = SHA512_DIGEST_LENGTH;
+    if (is384 == 1) {
+      digest_length = SHA384_DIGEST_LENGTH;
+    }
+
+    // STEP 1, get the HMAC-SHA512 hash from counter and key
+    uint8_t hash[digest_length];
+    HMAC_SHA512(key, key_length, in, 8, hash, is384);
```
Comment (joeycastillo) [inline movement/lib/TOTP/sha512.c:406]: Same issue as in TOTP_HMAC_SHA1.

File: movement/watch_faces/complication/totp_face.c [hunk 5]
```diff
@@ -6,24 +6,49 @@
 #include "TOTP.h"
 
 // Use https://cryptii.com/pipes/base32-to-hex to convert base32 to hex
-// Use https://totp.danhersam.com/ to generate test codes for verification
+// Use https://github.com/susam/mintotp to generate test codes for verification
 
-static const uint8_t num_keys = 2;
+static const uint8_t num_keys = 6;
```
Comment (joeycastillo) [inline movement/watch_faces/complication/totp_face.c:11]: No need to make this change yourself, but before merging I'm going to restore this to just two keys, and add some notes above and below signaling to folks that this is where they put their key data and what the different definitions mean.

**Review by joeycastillo — COMMENTED** [review]
Looks good overall! I just have one question about a hard-coded message length when hashing the keys; let me know what you think, and thanks!

File: movement/lib/TOTP/sha1.c (hunk 2 above)
Comment (EmilienCourt) [reply movement/lib/TOTP/sha1.c]: You are right :). This was fixed in https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/db2b82a5dc0e310489e966d1cff902568627d642.

File: movement/lib/TOTP/sha256.c (hunk 3 above)
Comment (EmilienCourt) [reply movement/lib/TOTP/sha256.c]: You are right :). This was fixed in https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/db2b82a5dc0e310489e966d1cff902568627d642.

File: movement/lib/TOTP/sha512.c (hunk 4 above)
Comment (EmilienCourt) [reply movement/lib/TOTP/sha512.c]: You are right :). This was fixed in https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/db2b82a5dc0e310489e966d1cff902568627d642.

File: movement/watch_faces/complication/totp_face.c (hunk 5 above)
Comment (EmilienCourt) [reply movement/watch_faces/complication/totp_face.c:11]: Alright, i'll let you do the modification then :)

**Comment by EmilienCourt** [comment]
Thanks for the feedback !

I'm still waiting to be able to test it on real hardware.

**Pull request merged**